test/e2e/windows/dns.go

```
@@ -37,58 +35,93 @@ var _ = SIGDescribe("[Feature:Windows] DNS", func() {
 	})
 
 	f := framework.NewDefaultFramework("dns")
-
 	ginkgo.It("should support configurable pod DNS servers", func() {
-		ginkgo.By("Preparing a test DNS service with injected DNS names...")
-		testInjectedIP := "1.1.1.1"
-		testSearchPath := "resolv.conf.local"
 
-		ginkgo.By("Creating a pod with dnsPolicy=None and customized dnsConfig...")
-		testUtilsPod := e2epod.NewAgnhostPod(f.Namespace.Name, "e2e-dns-utils", nil, nil, nil)
-		testUtilsPod.Spec.DNSPolicy = v1.DNSNone
-		testUtilsPod.Spec.DNSConfig = &v1.PodDNSConfig{
+		ginkgo.By("Getting the IP address of the internal Kubernetes service")
+
+		svc, err := f.ClientSet.CoreV1().Services("kube-system").Get(context.TODO(), "kube-dns", metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		ginkgo.By("Preparing a test DNS service with injected DNS names...")
+		// the default service IP will vary from cluster to cluster, but will always be present and is a good DNS test target
+		testInjectedIP := svc.Spec.ClusterIP
```

Since the default DNS for a pod is the `kube-dns` pod, if something isn't working with the custom DNS would these tests pass?

---

These tests will fail if there is no valid DNS service. I  figured thats actually a requirement - we can't test that we `support configurable pod DNS` unless we actually try to hit a named endpoint .

---

Any other test, which scans an implementation detail, wont actually measure the functionality, but rather, will just be an guess based on a heurisitc.

---

I am wondering on the use of the `kube-dns` as the endpoint though.  If the configuration is a noop for CNI's that don't implement the dns plugin then the default dns would be the cluster DNS (`kube-dns`) wouldn't it?  That would mean they would pass the test even though they didn't really configure the DNS?    I think we would need to use something like `1.1.1.1` to make sure custom DNS is working.

Agreed on actually hitting the endpoint. If the above is true then I think we should also keep the check that the DNS is actually set to what we expected and hit the endpoint (to make sure we are using the correct DNS).

---

oh, i see so you mean you could just 'get lucky' here, in cases where the dns isnt being overwritten

---

- Confirmed that no-op CNIs, like EKS, indeed cannot resolve `kubernetes`.  So were all good there.
- Also, just to be double careful, added a failure case, where 1.1.1.1 not being injected directly leads to failed test as well, that will weed out no-op DNS systems early on.
- So now we have both the 1.1.1.1 verification, as well as the other one.  BTW the old regex seemed to not be restrictive enough, so i just did a regular string search, more readable and more precise. 

- On EKS, this test now FAILS FAST (because it notices that 1.1.1.1 is not injected)
- On AKS w/ calico, the test passes .

---

> Confirmed that no-op CNIs, like EKS, indeed DO NOT by default resolve kubernetes on windows

Interesting. What is the default DNS configuration for EKS if not using kube-dns?  On azure with flannel using no DNS pod spec I get:

```
ipconfig /all
  DNS Servers . . . . . . . . . . . : 10.96.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       default.svc.cluster.local
                                       svc.cluster.local
                                       cluster.local

curl.exe -k https://kubernetes:443
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "forbidden: User \"system:anonymous\" cannot get path \"/\"",
  "reason": "Forbidden",
  "details": {

  },
  "code": 403
}
```

Where 10.96.0.10 is the kube-dns ip:

```
k get svc -n kube-system kube-dns              
NAME       TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   20h
```

---

interestinggg
in that case its goood we kept the 1.1.1.1 testcase

---

```PS C:\> ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : windows-server-iis-675f9744d7-dr2qv
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : omg.svc.cluster.local
                                       svc.cluster.local
                                       cluster.local

Ethernet adapter vEthernet (cid-5eae4daf98cf88ff695042b491c118892dcc853114f5353379c2b65c3e181d97):

   Connection-specific DNS Suffix  . : omg.svc.cluster.local
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #5
   Physical Address. . . . . . . . . : 00-15-5D-AC-88-C9
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::b131:bcb8:c3fa:96db%34(Preferred)
   IPv4 Address. . . . . . . . . . . : 192.168.22.170(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.224.0
   Default Gateway . . . . . . . . . : 192.168.0.1
   DNS Servers . . . . . . . . . . . : 10.100.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       omg.svc.cluster.local
                                       svc.cluster.local
                                       cluster.local```

thtas the default config of eks

---

maybe its because your running in the `default` namespaces which is why you resolve `kubernetews` 

otherwise, you wont be able to, hence why in general , the 2nd test is sufficient bc e2es allways run outside the default ns :)